src/dsl/dsl.jl

```diff
@@ -71,21 +72,77 @@ function address_from_expression(lhs)
 end
 
 function desugar_tildes(expr)
+    trace_ref = GlobalRef(@__MODULE__, Symbol("@trace"))
+    line_num = LineNumberNode(1, :none)
     MacroTools.postwalk(expr) do e
+        # Replace with globally referenced macrocalls
         if MacroTools.@capture(e, {*} ~ rhs_)
-            :(@trace($rhs))
+            Expr(:macrocall, trace_ref, line_num, rhs)
         elseif MacroTools.@capture(e, {addr_} ~ rhs_)
-            :(@trace($rhs, $(addr)))
+            Expr(:macrocall, trace_ref, line_num, rhs, addr)
         elseif MacroTools.@capture(e, lhs_ ~ rhs_)
-            addr_expr = address_from_expression(lhs)
-            :($lhs = @trace($rhs, $(addr_expr)))
+            addr = address_from_expression(lhs)
+            Expr(:(=), lhs, Expr(:macrocall, trace_ref, line_num, rhs, addr))
         else
             e
         end
     end
 end
 
-function parse_gen_function(ast, annotations)
+function resolve_gen_macros(expr, __module__)
+    MacroTools.postwalk(expr) do e
+        # Resolve Gen macros to globally referenced macrocalls
+        if (MacroTools.@capture(e, @namespace_.m_(args__)) &&
+            m in DSL_MACROS && __module__.eval(namespace) == @__MODULE__)
+            macro_ref = GlobalRef(@__MODULE__, m)
+            line_num = e.args[2]
+            Expr(:macrocall, macro_ref, line_num, args...)
+        elseif (MacroTools.@capture(e, @m_(args__)) &&
+                m in DSL_MACROS && isdefined(__module__, m) &&
+                getfield(__module__, m) == getfield(@__MODULE__, m))
+            macro_ref = GlobalRef(@__MODULE__, m)
+            line_num = e.args[2]
+            Expr(:macrocall, macro_ref, line_num, args...)
+        else
+            e
+        end
+    end
+end
+
+function extract_quoted_exprs(expr)
+    quoted_exprs = []
+    expr = MacroTools.prewalk(expr) do e
+        if MacroTools.@capture(e, :(quoted_))
+            push!(quoted_exprs, e)
+            Expr(:placeholder, length(quoted_exprs))
+        else
+            e
+        end
+    end
+    return expr, quoted_exprs
+end
+
+function insert_quoted_exprs(expr, quoted_exprs)
+    expr = MacroTools.prewalk(expr) do e
+        if MacroTools.@capture(e, p_placeholder)
+            idx = p.args[1]
+            quoted_exprs[idx]
+        else
+            e
+        end
+    end
+    return expr
+end
+
+function preprocess_body(expr, __module__)
```

Comment:
Could you add your explanation in https://github.com/probcomp/Gen.jl/pull/223#issuecomment-646420005 as a docstring here, or wherever you think it best fits?

Reply:
Just added explanatory comments to each line. Decided against docstrings for now -- we probably want to add docstrings consistently to all the parsing code at some later point.